[WIP] Add GZ3D vac code

CKrawczyk opened the pull request:
Adds code for GZ3D VAC integration.

Things left to do:
- [x] Ensure all custom class methods work as expected
- [x] Document how to use the class
- [x] Included docstrings on the class
- [x] Properly handle covariance when making stacked spectra
- [x] Create examples for using GZ3D masks
- [x] Add plotting methods to the GZ3D class for common plots

This pull request:
- [x] Has a title that summarises what is changing.
- [ ] Updates the documentation accordingly.
- [ ] Has unit tests & [code coverage](https://coveralls.io/github/sdss/marvin) is not adversely affected (within reason).
- [ ] Works with Python 2.7 and 3.6 (and ideally with Python 3.7).
- [ ] Updates the [CHANGELOG](https://github.com/sdss/marvin/blob/master/CHANGELOG.rst).
- [ ] Removes more lines of code than it adds.
- [ ] If relevant, adds a new entry to the [What's new?](https://github.com/sdss/marvin/blob/master/docs/sphinx/whats-new.rst) page.

havok2063 commented on python/marvin/contrib/vacs/galaxyzoo3d.py:
```diff
@@ -0,0 +1,864 @@
+        # get the subject id from the filename
+        self.subject_id = filename.split('/')[-1].split('_')[-1].split('.')[0]
+        # read in the fits file
+        self.hdulist = fits.open(filename)
```
I wonder if you make this a contextmanager, i.e. `with fits.open(filename) as self.hdulist` it will handle edge cases where something breaks in the code, at least the file will be closed.  Then you could remove the explicit `close` method.

CKrawczyk replied:
good idea, I have updated the code.

havok2063 commented on python/marvin/contrib/vacs/galaxyzoo3d.py:
```diff
@@ -0,0 +1,864 @@
+        else:
+            self.sf_mask = sf_mask_nii & sf_mask_sii
+
+    def bpt_in_mask(self, mask_name, bpt_name, factor=1.2):
```
docstring..

havok2063 reviewed (changes requested):
Is this still a WIP?  I would add docstrings everywhere there isn't it.  Complete ones for exposed methods, single line ones for hidden methods.  Overall there's a lot going on here.  Which are the key methods that you want exposed to the user?  I'm guessing the `plot` methods.  Are there methods that return useful data to the user?  I would recommend making everything that is really unnecessary to the user hidden.

CKrawczyk commented:
The WIP is exactly there because of the lack of doc string.  I will try to get to this soon, but other projects are eating my time at the moment.  Also how important are unit tests for VAC code?

CKrawczyk commented:
I just added doc strings to the private function, just need to add them to the plotting functions.  While working on this code I created an example notebook that shows how to use it.  Is this something that can be used in marvin's docs?  If so where should it go and what format should it be?

https://gist.github.com/CKrawczyk/191897d179464a8270cda2d055a9f2a0

havok2063 commented:
Thanks for adding the docstrings!    Regarding tests, it's not super important.  Our test suite is a bit clunky, out-of-date, and broken at the moment.  That being said, if there were some tests you wanted to include as a sanity check, they could go in `tests/tools/test_vacs.py`.   But the tests might not run for you out of the box due to how our suite is setup, but I can either run them myself or tell you how to run them.  

The notebook looks great!  We can definitely put it in place.  It should go in the `docs/sphinx/tutorials/notebooks/` folder, and I'd follow the format of the other VAC notebook, `VAC_[notebook_name].ipynb`.  And then add new lines in the `docs/sphinx/tutorials/index.rst` file. 

Let me know when you think this is ready to be merged.  I'm aiming to have all the VAC PRs merged in the next few weeks so I can begin testing and finalization for DR17.

havok2063 commented:
@CKrawczyk Are you ready for this to be merged?  I need to merge this as soon as possible so I can start finalizing the release of Marvin for DR17 in less than a month.

CKrawczyk commented on python/marvin/contrib/vacs/galaxyzoo3d.py:
```diff
@@ -0,0 +1,864 @@
+    URL: <look this up>
+
+    Description: <look this up>
+
```
What URL should go here?

CKrawczyk commented:
@havok2063 I think this is ready to go, I just need to know what URL you want to be included in the docstring for the `GZ3DVAC` class?

I have done a test build of the docs locally to make sure the example notebook shows up as expected.

For good measure, I also rebased against the latest `vacs_dr17` branch and fixed up any merge conflicts.

CKrawczyk commented:
The rebase basically just replayed my commits on top of any there were made to the base branch, any time a conflict was found I resolved it (just in the `index.rst` of the notebooks doc page).  

This will merge into the base branch as if all my commits were made from a fresh branch made today rather than one made several months ago.  All in all the git history will look "cleaner" than if I merged the base branch into this branch only for it to be merged back into the base branch.

I will double-check with Karen if we have a VAC URL yet.

havok2063 commented:
Sounds good to me.
